lib/Dialect/Calyx/CalyxOps.cpp

```diff
@@ -1279,6 +1282,248 @@ SmallVector<DictionaryAttr> InstanceOp::portAttributes() {
   return portAttributes;
 }
 
+//===----------------------------------------------------------------------===//
+// PrimitiveOp
+//===----------------------------------------------------------------------===//
+
+/// Lookup the component for the symbol. This returns null on
+/// invalid IR.
+hw::HWModuleExternOp PrimitiveOp::getReferencedPrimitive() {
+  auto program = (*this)->getParentOfType<ProgramOp>();
+  if (!program)
+    return nullptr;
+
+  return program.lookupSymbol<hw::HWModuleExternOp>(primitiveName());
+}
+
+/// Verifies the port information in comparison with the referenced component
+/// of an instance. This helper function avoids conducting a lookup for the
+/// referenced component twice.
+static LogicalResult verifyPrimitiveOpType(PrimitiveOp instance,
+                                          hw::HWModuleExternOp referencedPrimitive) {
+  auto program = instance->getParentOfType<ProgramOp>();
+  StringRef entryPointName = program.entryPointName();
+  if (instance.primitiveName() == entryPointName)
+    return instance.emitOpError()
+           << "cannot reference the entry-point component: '" << entryPointName
+           << "'.";
+
+  // Verify the instance result ports with those of its referenced component.
+  SmallVector<hw::PortInfo> primitivePorts = referencedPrimitive.getAllPorts();
+  size_t numPorts = primitivePorts.size();
+
+  size_t numResults = instance.getNumResults();
+  if (numResults != numPorts)
+    return instance.emitOpError()
+           << "has a wrong number of results; expected: " << numPorts
+           << " but got " << numResults;
+
+  for (size_t i = 0; i != numResults; ++i) {
+    auto resultType = instance.getResult(i).getType();
+    auto expectedType = primitivePorts[i].type;
+    auto replacedType = hw::evaluateParametricType(instance.getLoc(), instance.parametersAttr(), expectedType);
+    if (resultType == replacedType)
+      continue;
+    return instance.emitOpError()
+           << "result type for " << primitivePorts[i].name << " must be "
+           << expectedType << ", but got " << resultType;
+  }
+  return success();
+}
+
+LogicalResult PrimitiveOp::verifySymbolUses(SymbolTableCollection &symbolTable) {
```

Comment:
Looks like this function is virtually the same as `InstanceOp::verifySymbolUses` - please merge them, if possible, to make things more DRY.

Reply:
Although these functions are quite similar, the error messages have to be different depending on whether it is a primitive or a component. I can't come up with a good way to merge these functions with less code than is currently required.